executor/new_executor_xapi.go

```diff
@@ -228,70 +223,89 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
 		}
-		log.Debugf("[TIME_INDEX_SCAN] time: %v handles: %d", time.Now().Sub(startTs), len(handles))
-		e.buildTableTasks(handles)
-		limitCount := int64(-1)
-		if e.indexPlan.LimitCount != nil {
-			limitCount = *e.indexPlan.LimitCount
-		}
-		concurrency := 2
-		if limitCount == -1 {
-			concurrency = len(e.tasks)
-		} else if limitCount > int64(BaseLookupTableTaskSize) {
-			concurrency = int(limitCount/int64(BaseLookupTableTaskSize) + 1)
-		}
-		log.Debugf("[TIME_INDEX_TABLE_CONCURRENT_SCAN] start %d workers", concurrency)
-		e.runTableTasks(concurrency)
+		idxResult.Fetch()
+
+		// Use a background goroutine to fetch index, put the result in e.tasks.
+		// e.tasks serves as a pipeline, so fetch index and get table data would
+		// run concurrency.
+		e.tasks = make(chan *lookupTableTask, 50)
+		go e.fetchHandles(idxResult, e.tasks)
 	}
+
 	for {
-		if e.taskCursor >= len(e.tasks) {
-			return nil, nil
-		}
-		task := e.tasks[e.taskCursor]
-		if !task.done {
-			startTs := time.Now()
-			err := <-task.doneCh
-			if err != nil {
-				return nil, errors.Trace(err)
+		if e.taskCurr == nil {
+			taskCurr, ok := <-e.tasks
+			if !ok {
+				log.Debugf("[TIME_INDEX_TABLE_SCAN] time: %v", time.Now().Sub(startTs))
+				return nil, e.tasksErr
 			}
-			task.done = true
-			log.Debugf("[TIME_INDEX_TABLE_SCAN] time: %v handles: %d", time.Now().Sub(startTs), len(task.handles))
+			e.taskCurr = taskCurr
 		}
-		if task.cursor < len(task.rows) {
-			row := task.rows[task.cursor]
-			task.cursor++
-			return row, nil
+
+		row, err := e.taskCurr.getRow()
+		if err != nil || row != nil {
+			return row, errors.Trace(err)
 		}
-		e.taskCursor++
+		e.taskCurr = nil
 	}
 }
 
-// fetchHandle fetches all handles from the index.
-// This should be optimized to fetch handles in batch.
-func (e *NewXSelectIndexExec) fetchHandles() ([]int64, error) {
-	idxResult, err := e.doIndexRequest()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	idxResult.Fetch()
-	handles, err := extractHandlesFromIndexResult(idxResult)
-	if err != nil {
-		return nil, errors.Trace(err)
+const concurrencyLimit int = 30
+
+// addWorker add a worker for lookupTableTask.
+// It's not thread-safe and should be called in fetchHandles goroutine only.
+func addWorker(e *NewXSelectIndexExec, ch chan *lookupTableTask, concurrency *int) {
+	if *concurrency <= concurrencyLimit {
+		go e.pickAndExecTask(ch)
+		*concurrency = *concurrency + 1
 	}
-	if !e.indexPlan.OutOfOrder {
-		// Save the index order.
-		e.indexOrder = make(map[int64]int)
-		for i, h := range handles {
-			e.indexOrder[h] = i
+}
+
+func (e *NewXSelectIndexExec) fetchHandles(idxResult xapi.SelectResult, ch chan<- *lookupTableTask) {
+	defer close(ch)
+
+	workCh := make(chan *lookupTableTask, 1)
+	defer close(workCh)
+
+	var concurrency int
+	addWorker(e, workCh, &concurrency)
+
+	totalHandles := 0
+	startTs := time.Now()
+	for {
+		handles, err := extractHandlesFromIndexResult(idxResult)
+		if err != nil || handles == nil {
```

Comment:
Is it possible that a partial result returns nil handles, but there still other handles in other partial result to be extracted?